dll/win32/kernel32/kernel32_vista/GetFinalPathNameByHandle.c

```diff
@@ -0,0 +1,298 @@
+                           NULL);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("NtQueryObject failed: %lx\n", Status);
+        BaseSetLastNTError(Status);
+        return 0;
+    }
+
+    /* If the NT path was requested, we return the name as it is */
+    if (VolumeType == VOLUME_NAME_NT)
+    {
+        /* Calculate the final length and validate the buffer size */
+        FinalLength = NameInfo->Name.Length / sizeof(WCHAR);
+        if (cchFilePath < FinalLength + 1)
+        {
+            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+            return FinalLength + 1;
+        }
+
+        /* Copy the name to the caller's buffer */
+        RtlCopyMemory(lpszFilePath, NameInfo->Name.Buffer, NameInfo->Name.Length);
+        lpszFilePath[NameInfo->Name.Length / sizeof(WCHAR)] = UNICODE_NULL;
+
+        /* Return the length of the name */
+        return FinalLength;
+    }
+
+    /* For everything else we need to split the path */
+    PathSplit = wcschr(&NameInfo->Name.Buffer[sizeof("\\Device\\") - 1], L'\\');
+    if (PathSplit == NULL)
+    {
+        DPRINT1("Invalid object name: %wZ\n", &NameInfo->Name);
+        SetLastError(ERROR_INVALID_HANDLE);
+        return 0;
+    }
+
+    DeviceName.Buffer = NameInfo->Name.Buffer;
+    DeviceName.Length = (USHORT)(PathSplit - NameInfo->Name.Buffer) * sizeof(WCHAR);
+    DeviceName.MaximumLength = DeviceName.Length;
+
+    RelativePath.Buffer = PathSplit + 1;
+    RelativePath.Length = NameInfo->Name.Length - DeviceName.Length - sizeof(WCHAR);
+    RelativePath.MaximumLength = RelativePath.Length;
+
+    /* Query the DOS volume name */
+    RtlInitEmptyUnicodeString(&VolumeName, VolumeBuffer, sizeof(VolumeBuffer));
+    Status = QueryDosVolumeNameForNtDeviceName(&DeviceName, &VolumeName);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("QueryDosVolumeNameForNtDeviceName failed: %lx\n", Status);
+        BaseSetLastNTError(Status);
+        return 0;
+    }
+
+    if (VolumeType == VOLUME_NAME_DOS)
+    {
+        /* Calculate the final length and validate the buffer size */
+        FinalLength = ((VolumeName.Length + RelativePath.Length) / sizeof(WCHAR));
+        if (cchFilePath < FinalLength + 1)
+        {
+            SetLastError(ERROR_NOT_ENOUGH_MEMORY);
+            return FinalLength + 1;
+        }
+
+        /* Construct the final name */
+        RtlCopyMemory(lpszFilePath, VolumeName.Buffer, VolumeName.Length);
+        RtlCopyMemory(lpszFilePath + (VolumeName.Length / sizeof(WCHAR)),
+                      RelativePath.Buffer,
+                      RelativePath.Length);
+        lpszFilePath[FinalLength] = UNICODE_NULL;
+
+        /* Return the length of the name */
+        return FinalLength;
+    }
+    else if (VolumeType == VOLUME_NAME_GUID)
+    {
+        /* Query the GUID volume name */
+        Success = GetVolumeNameForVolumeMountPointW(VolumeBuffer,
+                                                    VolumeBuffer,
```

Comment:
Is it valid to invoke the function on the same buffer for both the input and the output parameter? It may appear to work but usually such a thing could be undefined behaviour (dependent on the exact implementation of GetVolumeNameForVolumeMountPointW).